rts/Game/UI/KeyBindings.cpp

```diff
@@ -311,82 +310,115 @@ void CKeyBindings::Kill()
 
 	// We find the position that partitions the vector
-	auto anyIt = std::find_if(actionList.begin(), actionList.end(), [](const Action& a) {
-		return a.keyChain.back().AnyMod();
-	});
+	auto getAnyIt = [](const ActionList & list) {
+		return std::find_if(std::begin(list), std::end(list), [](const Action& a) {
+			return a.keyChain.back().AnyMod();
+		});
+	};
+
+	const auto aAnyIt = getAnyIt(actionListA);
+	const auto bAnyIt = getAnyIt(actionListB);
+
+	// Only add non-Any items from list A
+	merged.insert(std::end(merged), std::begin(actionListA), aAnyIt);
+	size_t aEndId = merged.size();
+
+	// Only add non-Any items from list B
+	// - If there are duplicates with >= bindingId, don't add them
+	// - If there are duplicates with < bindingId, remove the A item
+	auto addItemsFromB = [&merged](auto bBeginIt, auto bEndIt, size_t aBeginId, size_t aEndId) {
+		for (auto bIt = bBeginIt; bIt < bEndIt; ++bIt) {
+			const auto & aB = *bIt;
+			bool toAdd = true;
+			for (size_t a = aBeginId; a < aEndId; ++a) {
+				// B is a duplicate...
+				if (aB.line == merged[a].line) {
+					// ...with a higher id, so do not add it.
+					if (aB.bindingIndex >= merged[a].bindingIndex) {
+						toAdd = false;
+					// ...with a lower id, so remove its A equivalent.
+					} else {
+						merged.erase(std::next(std::begin(merged), a));
+						--aEndId;
+					}
+					break;
+				}
+			}
+			if (toAdd) merged.push_back(aB);
+		}
+		return aEndId;
+	};
+
+	aEndId = addItemsFromB(std::begin(actionListB), bAnyIt, 0, aEndId);
+
+	// Finally merge the two non-Any lists
+	std::inplace_merge(std::begin(merged), std::next(std::begin(merged), aEndId), std::end(merged), compareActionByTriggerOrder);
+	size_t aBeginId = merged.size();
 
-	// We remove all duplicates without anyMod
-	anyIt = actionList.erase(std::remove_if(actionList.begin(), anyIt, lineMatcher), anyIt);
+	// Add the Any A part
+	merged.insert(std::end(merged), aAnyIt, std::end(actionListA));
+	aEndId = merged.size();
 
-	// We clear the duplicate memory
-	seen.clear();
+	// Add the Any B part
+	aEndId = addItemsFromB(bAnyIt, std::end(actionListB), aBeginId, aEndId);
 
-	// We remove all duplicates with anyMod
-	actionList.erase(std::remove_if(anyIt, actionList.end(), lineMatcher), actionList.end());
+	// Merge the two any parts.
+	std::inplace_merge(std::next(std::begin(merged), aBeginId), std::next(std::begin(merged), aEndId), std::end(merged), compareActionByTriggerOrder);
 
-	return actionList;
+	return merged;
 }
 
 
-CKeyBindings::ActionList CKeyBindings::MergeActionLists(const ActionList& actionListA, const ActionList& actionListB, ActionComparison compare = compareActionByTriggerOrder)
+CKeyBindings::ActionList CKeyBindings::GetActionList() const
 {
-	if (actionListA.empty())
-		return actionListB;
-	if (actionListB.empty())
-		return actionListA;
-
 	ActionList merged;
-	merged.reserve(actionListA.size() + actionListB.size());
+	//merged.reserve(sizeof(defaultBindings) / sizeof(defaultBindings[0]));
```

Comment:
There might be a better heuristic to better reserve entries, it's pretty safe to assume the list of bound actions will be higher than the default bindings count for the majority of cases